### PR TITLE
Multiselect for advanced options

### PR DESCRIPTION
**Pull request by MaximilianKrambach**
Expanding on the new dialog from @emanuelschuetze (from PR #4037), and (re-)using bits of PR #3963 (for retrieving categories, motions, etc...)

### CODE REVIEW

**Comment by MaximilianKrambach** [comment]
@emanuelschuetze What about multiple choices as in user.groups:

Should we
a) add
b) replace

the selected groups to all users?

(example: User 1 is Delegate, User 2 staff. I select both and set groups to "Administration". Should they have administration *added*, or should they then only be administration)?

**Comment by MaximilianKrambach** [comment]
@emanuelschuetze 
I am unsure, too, what a "recommendation" for a motion is.

tsiegleauq pointed out a problem with big lists: The mat-dialog does not scroll by default, as it is not supposed to have so many entries (thinking of 'all tags').

For my previous question, I would suggest two buttons: `Add` and `Overwrite` (in place of the `OK` button).

**Comment by MaximilianKrambach** [comment]
(Another place of confusion: Should I open a list/dropdown of *all users* for 'set submitter') ?

**Comment by emanuelschuetze** [comment]
> @emanuelschuetze What about multiple choices as in user.groups:
> 
> Should we
> a) add
> b) replace
> 
> the selected groups to all users?

Please add/remove selected groups - NOT replace. See also version 2.x  (demo.openslides.org)

**Comment by emanuelschuetze** [comment]
> I am unsure, too, what a "recommendation" for a motion is.

Same as 'set state' action: List all available recommendations (add recommender in motion settings before). Set selected recommendation for all selected motions.

> The mat-dialog does not scroll by default, 

I think we should make the dialog scrollable (categories, tags, states, recommendations etc should fit in a scrollable list).

> (Another place of confusion: Should I open a list/dropdown of all users for 'set submitter') ?

Yes, for 'set submitters' use a dropdown list.

> I would suggest two buttons: Add and Overwrite (in place of the OK button).

Use 'add' and 'remove' button - and additional a 'Cancel' button.

**Comment by MaximilianKrambach** [comment]
>>I am unsure, too, what a "recommendation" for a motion is.

>Same as 'set state' action: List all available recommendations (add recommender in motion settings before). Set selected recommendation for all selected motions.

This does not make it clear what an 'recommendation' is in terms of models. Is it a 'change recommendation'? Or something else? And no, I didn't find it in 2.3

**Comment by tsiegleauq** [comment]
@MaximilianKrambach 
recommendations are, essentially, workflow states.

See [ViewMotion](https://github.com/OpenSlides/OpenSlides/blob/master/client/src/app/site/motions/models/view-motion.ts#L159)

File: client/src/app/site/motions/components/motion-list/motion-list.component.html [hunk 1]
```diff
@@ -139,25 +132,62 @@ <h2 translate>Motions</h2>
...
+            </button>
         </div>
-        <button mat-menu-item (click)="csvExportMotionList()">
+        <mat-divider></mat-divider>
+        <button mat-menu-item (click)="csvExportMotionList(); toggleMultiSelect()">
             <mat-icon>archive</mat-icon>
             <span translate>Export as CSV</span>
         </button>
+        <mat-divider></mat-divider>
+        <button mat-menu-item (click)="multiselectService.delete(selectedRows); toggleMultiSelect()">
```
Comment (tsiegleauq) [inline client/src/app/site/motions/components/motion-list/motion-list.component.html]: class="red-warning-text"

File: client/src/app/site/motions/models/view-workflow.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    public get states() : WorkflowState[] {
+        return this.workflow ? this.workflow.states : null;
+    }
+
+    public get first_state(): number {
+        return this.workflow ? this.workflow.first_state : null;
+    }
+
+
```
Comment (tsiegleauq) [inline client/src/app/site/motions/models/view-workflow.ts]: line too much

File: client/src/app/site/users/components/user-list/user-list.component.html [hunk 3]
```diff
@@ -123,43 +131,40 @@ <h2 translate>Participants</h2>
...
-                <span translate>Send invitations</span>
+                <span translate>Send invitation emails</span>
             </button>
-
+            <button mat-menu-item (click)="resetPasswordsSelected()">
+                <mat-icon>vpn_key</mat-icon>
+                <span translate>Generate new passwords</span>
+            </button>
+            <mat-divider></mat-divider>
             <button mat-menu-item (click)="deleteSelected()">
```
Comment (tsiegleauq) [inline client/src/app/site/users/components/user-list/user-list.component.html]: [value]="configItem.value"

File: client/src/app/site/users/services/user-repository.service.ts [hunk 4]
```diff
@@ -84,4 +89,70 @@ export class UserRepositoryService extends BaseRepository<ViewUser, User> {
...
+    public getRandomPassword(length: number = 8): string {
+        let pw = '';
+        const characters = 'abcdefghijkmnpqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789';
+        for (let i = 0; i < length; i++) {
+            pw += characters.charAt(Math.floor(Math.random() * characters.length));
+        }
+        return pw;
+    }
+
+    public async resetPassword(user: ViewUser, password: string): Promise<void> {
```
Comment (tsiegleauq) [inline client/src/app/site/users/services/user-repository.service.ts]: todo: canmanage or something that only admins might do

File: client/src/app/shared/components/choice-dialog/choice-dialog.component.html [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<h2 mat-dialog-title>{{ data.title | translate }}</h2>
+<div class="scrollmenu">
```
[the remark this answers is not in the document]
Comment (FinnStutzenstein) [reply client/src/app/shared/components/choice-dialog/choice-dialog.component.html]: @MaximilianKrambach

File: client/src/app/shared/components/choice-dialog/choice-dialog.component.scss [hunk 6]
```diff
@@ -0,0 +1,15 @@
+mat-radio-group {
+    display: inline-flex;
+    flex-direction: column;
+
+    mat-radio-button {
+        margin: 5px;
+    }
+}
+
+.scrollmenu {
```
[the remark this answers is not in the document]
Comment (FinnStutzenstein) [reply client/src/app/shared/components/choice-dialog/choice-dialog.component.scss]: @MaximilianKrambach

File: client/src/app/site/users/components/user-list/user-list.component.html (hunk 3 above)
Comment (FinnStutzenstein) [reply client/src/app/site/users/components/user-list/user-list.component.html]: ? I guess you mean the css-class from above.

File: client/src/app/site/users/services/user-repository.service.ts (hunk 4 above)
Comment (FinnStutzenstein) [reply client/src/app/site/users/services/user-repository.service.ts]: it is protected with *osPerms in user-list.component.html line 117

File: client/src/app/site/users/components/user-list/user-list.component.html (hunk 3 above)
Comment (tsiegleauq) [reply client/src/app/site/users/components/user-list/user-list.component.html]: Yes. :-D

File: client/src/app/shared/components/choice-dialog/choice-dialog.component.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (MaximilianKrambach) [reply client/src/app/shared/components/choice-dialog/choice-dialog.component.html]: I think there is a solution now which works. I just hope that I'll still be able to push into my own PR...

**Comment by FinnStutzenstein** [comment]
@tsiegleauq Please check new commits.
Until now where submitters just could be created, everything was fine. Now I needed to handle them (aggregate, edit etc) and the "hack" with the `submitters_id` in the motion model was conflicting.
It was introduced, because on motion creation, you can send an array of user_ids to be submitters. If you work with motions from the DataStore, this attribute is undefined. Getter and setters in the `ViewMotion` were dealing with these attributes in ways, that didn't work here anymore. If you edit a motion, the handling of submitters is different then during creation.

I removed the "hacks" for motion creations and make a new class and a new model extending `Motion` and `ViewMotion` to handle the special submitter things. This needs to be checked.

I added all new multiselect routes. Everything should work so far. @emanuelschuetze Please verify. Currently, I try to fix the trvis error.

**Comment by MaximilianKrambach** [comment]
@FinnStutzenstein I am confused about the need of 'multiselectWrapper', which appeared quite late in the PR process. What is it for (except for 'exit multiSelect', which is not always wanted)?

**Comment by FinnStutzenstein** [comment]
It's for catching exceptions that might be raised by some multiselect actions. All actions (respective the methods in the multiselect service) return promises. The wrapper appends a .then to each action. On error, the error is shown; on success the multiselect is closed

**Pull request merged**